src/systems/sound-effects-system.js

```diff
@@ -0,0 +1,245 @@
+/* global fetch THREE AFRAME */
+import TICK from "../assets/sfx/tick.mp3";
+import TELEPORT_ARC from "../assets/sfx/teleportArc.mp3";
+import QUICK_TURN from "../assets/sfx/quickTurn.mp3";
+import TAP_MELLOW from "../assets/sfx/tap_mellow.mp3";
+import PEN_SPAWN from "../assets/sfx/PenSpawn.mp3";
+import PEN_DRAW from "../assets/sfx/PenDraw1.mp3";
+import CAMERA_SNAPSHOT from "../assets/sfx/PicSnapHey.mp3";
+import WELCOME from "../assets/sfx/welcome.mp3";
+import QUACK from "../assets/sfx/quack.mp3";
+import SPECIAL_QUACK from "../assets/sfx/specialquack.mp3";
+import POP from "../assets/sfx/pop.mp3";
+import FREEZE from "../assets/sfx/Eb_blip.mp3";
+import TACK from "../assets/sfx/tack.mp3";
+import MEDIA_LOADED from "../assets/sfx/A_bendUp.mp3";
+import MEDIA_LOADING from "../assets/sfx/suspense.mp3";
+import { paths } from "./userinput/paths";
+
+function getBuffer(url, context) {
+  return fetch(url)
+    .then(r => r.arrayBuffer())
+    .then(arrayBuffer => context.decodeAudioData(arrayBuffer));
+}
+
+function playSound(buffer, context) {
+  // The nodes are very inexpensive to create, according to
+  // https://developer.mozilla.org/en-US/docs/Web/API/AudioBufferSourceNode
+  const source = context.createBufferSource();
+  source.buffer = buffer;
+  source.connect(context.destination);
+  source.start();
+}
+
+function playSoundLooped(buffer, context) {
+  const source = context.createBufferSource();
+  source.buffer = buffer;
+  source.loop = true;
+  source.connect(context.destination);
+  source.start();
+  return source;
+}
+
+function copy(current, prev) {
+  prev.held = current.held;
+  prev.hovered = current.hovered;
+}
+
+function isUI(el) {
+  return (
+    el && el.components.tags && (el.components.tags.data.singleActionButton || el.components.tags.data.holdableButton)
+  );
+}
+
+export class SoundEffectsSystem {
+  constructor(scene) {
+    this.teleporters = {
+      leftHand: null,
+      rightHand: null,
+      rightRemote: null
+    };
+    this.prevInteractionState = {
+      leftHand: {
+        hovered: null,
+        held: null
+      },
+      rightHand: {
+        hovered: null,
+        held: null
+      },
+      rightRemote: {
+        hovered: null,
+        held: null
+      }
+    };
+    this.teleporterState = {
+      leftHand: {
+        teleporting: false,
+        teleportArcSource: null
+      },
+      rightHand: {
+        teleporting: false,
+        teleportArcSource: null
+      },
+      rightRemote: {
+        teleporting: false,
+        teleportArcSource: null
+      }
+    };
+    this.ctx = THREE.AudioContext.getContext();
+    this.pendingEffects = [];
+    this.soundFor = new Map();
+    this.sounds = {};
+    getBuffer(TICK, this.ctx).then(buffer => {
```

Comment:
All these getBuffer/then blocks are a lot of boilerplate code. Seems like after getting rid of `this.sounds` it should be possible to rework this into a `Promise.all` and a loop. This would also help get rid of the kind of ugly `shouldTick`.